internal/controller/instaslice_controller.go

```diff
@@ -105,61 +119,86 @@ func (r *InstasliceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusCreating {
+						return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
+					}
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusCreated || allocation.Allocationstatus == inferencev1alpha1.AllocationStatusUngated {
+						resultDeleting, errInDeleting := r.setInstasliceAllocationToDeleting(ctx, instaslice.Name, string(pod.UID), allocation)
+						if errInDeleting != nil {
+							return resultDeleting, nil
+						}
+						// return and rely on daemonset to se allocation status to created
+						// this will cause podmap function to wakeup pod and perform clean up
+						return ctrl.Result{}, nil
+					}
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusDeleted {
+						resultRemove, errInRemove := r.removeInstasliceAllocation(ctx, instaslice.Name, allocation)
+						if errInRemove != nil {
+							return resultRemove, nil
+						}
+						// requeue for the finalizer to be removed
+						return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
+					}
 				}
-
 			}
 		}
 		// pod can be terminated without any allocation
-		if controllerutil.RemoveFinalizer(pod, finalizerName) {
+		if allocationNotFound && controllerutil.RemoveFinalizer(pod, finalizerName) {
 			if err := r.Update(ctx, pod); err != nil {
 				log.FromContext(ctx).Error(err, "unable to update removal of finalizer, retrying")
 				// requeing immediately as the finalizer removal gets lost
 				return ctrl.Result{Requeue: true}, nil
 			}
-			log.FromContext(ctx).Info("finalizer deleted")
+			log.FromContext(ctx).Info("finalizer deleted for failed for ", "pod", pod.Name)
 		}
 		return ctrl.Result{}, nil
 	}
 
 	// pod is completed move allocation to deleting state and return
 	if pod.Status.Phase == v1.PodSucceeded && controllerutil.ContainsFinalizer(pod, finalizerName) {
+		allocationNotFound := true
 		for _, instaslice := range instasliceList.Items {
 			for _, allocation := range instaslice.Spec.Allocations {
-				found, result, err := r.processInstasliceAllocation(ctx, instaslice.Name, string(pod.UID), allocation, req)
-				if err != nil {
-					return result, err
-				}
-				if found {
-					return ctrl.Result{}, nil
+				if allocation.PodUUID == string(pod.UID) {
+					allocationNotFound = false
+					if allocation.Allocationstatus != inferencev1alpha1.AllocationStatusDeleted {
+						result, err := r.setInstasliceAllocationToDeleting(ctx, instaslice.Name, string(pod.UID), allocation)
+						if err != nil {
+							return result, err
+						}
+						// return and rely on daemonset to se allocation status to created
+						// this will cause podmap function to wakeup pod and perform clean up
+						return ctrl.Result{}, nil
+					}
+
+					if allocation.Allocationstatus == inferencev1alpha1.AllocationStatusDeleted {
+						result, err := r.removeInstasliceAllocation(ctx, instaslice.Name, allocation)
+						if err != nil {
+							return result, nil
```

Comment:
This is strange... err is not returned, and line 188 is retrying to requeue.

Reply:
states are different